Custom Import Order Check, updated third-party package property's defaul...

alexkravin opened the pull request:
...t value, issue #515

According to #515 

Modified the default value of <b>thirdPartyPackageRegExp</b> from empty string (^$) to either empty or not string (^.*$), so now user is able to not specifying this value as it's said in docs:
```
THIRD_PARTY_PACKAGE group. This group sets ordering of third party imports. Third party imports are all imports except STATIC, SAME_PACKAGE(n) and STANDARD_JAVA_PACKAGE.
```

Added corresponding UT and input.

romani commented:
While I am thinking about that issue and update, please generate report over guava (version 17.0) for that Check and lets compare it with report after update.
config, report is here - http://checkstyle.sourceforge.net/google_style.html

please also launch test of - https://github.com/checkstyle/google-style-config-test

alexkravin commented:
launched test of https://github.com/checkstyle/google-style-config-test
- No problems were occurred.

Generated 2 reports on guava project (but on version 19-snapshot, because I didn't find version 17 in branches of guava project):
1 - With standard configuration as here - https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml#L154

```xml
<module name="CustomImportOrder">
            <property name="thirdPartyPackageRegExp" value=".*"/>
            <property name="specialImportsRegExp" value="com.google"/>
            <property name="sortImportsInGroupAlphabetically" value="true"/>
            <property name="customImportOrderRules" value="STATIC###SPECIAL_IMPORTS###THIRD_PARTY_PACKAGE###STANDARD_JAVA_PACKAGE"/>
        </module>
```
Report: http://alexkravin.github.io/CustomImportOrder/former/guava/checkstyle.html

2 - Without specifying property <b>thirdPartyPackageRegExp</b> - with its default value
Report: http://alexkravin.github.io/CustomImportOrder/guava/checkstyle.html

These reports are the same.

romani commented:
@alexkravin , thanks a lot for testing, but we need to do a little bit more.

1) I agree with your fix to default value to ".*" (but not a "^.*$")

2) please update Javadoc to look nice with clear distinction of code and text 
http://checkstyle.sourceforge.net/apidocs/index.html?com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.html

3) Please add example to javadoc and xdoc of 
SAME_PACKAGE(3)###
THIRD_PARTY_PACKAGE(org.*)###
STATIC###
THIRD_PARTY_PACKAGE(com.*)###
SPECIAL_IMPORTS(android.*)

4) Please recheck that there is UT for case that I described in point 3.

5) Please recheck how SPECIAL_IMPORTS is affecting THIRD_PARTY_PACKAGE, and update javadoc/xdoc accordingly.

5) we need to update google_style.xml to remove settings for thirparty as it now default.

alexkravin commented:
1) default value is <b>.</b> now.

2) Updated with tag &lt;code&gt; where it is needed.

3) Done.

4) There was no such UT, added one and corresponding input, updated one line in finishTree() method as Check couldn't work normally if groups in custom order rules are not single. As it is in example there's repeating of THIRD_PARTY_PACKAGE

5) I think you meant problem I'm pointing on in previous paragraph. Example in javadoc/xdoc illustrates the possibility of such groups specifying.

6) Updated google_checks.xml

romani commented:
1) please make a default as ".\*" , sorry forget about that formatting nunace at github.

2) please create Input that cover a reported use case, it might looks like duplication of more complicated input file, but please do.

3) 
> I think you meant problem I'm pointing on in previous paragraph. 

I was not exact in my description, Please review and clarify how SPECIAL_PACKAGE is affecting THIRD_PARTY_PACKAGE. 

> THIRD_PARTY_PACKAGE group. This group sets ordering of third party imports. 
Third party imports are all imports except STATIC, SAME_PACKAGE(n) 
and STANDARD_JAVA_PACKAGE.

is there SPECIAL_PACKAGE is missed or it was by design. If is intensionally missed , I think we have design problem.

4) 
> As it is in example there's repeating of THIRD_PARTY_PACKAGE

Hmm, sorry , I mixedup that Check with other Check where regexp is defined inside macros.
So , we can not use two macroses THIRD_PARTY_PACKAGE, as it will not not clear what imports should go to fierst place nad what should go to second place (for regexp "org.*|android.*").

alexkravin commented:
1) Done

2) Was already done: https://github.com/alexkravin/checkstyle/blob/5d041a11ff78d11b31efcd6ff3b16601e3204e04/src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheckTest.java#L272

3) I think SPECIAL_IMPORTS were missed because previous author just forgotten about it, updated method matchesThirdPartyPackage(): https://github.com/alexkravin/checkstyle/blob/5d041a11ff78d11b31efcd6ff3b16601e3204e04/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java#L552

to follow updated docs: 
```
 * THIRD_PARTY_PACKAGE group. This group sets ordering of third party imports.
 * Third party imports are all imports except STATIC,
 * SAME_PACKAGE(n), STANDARD_JAVA_PACKAGE and SPECIAL_IMPORTS.
```

I think there is no need to get rid of specifying third_party_package_regex because it adds flexibility for uses if he wants it, e.g.:
User is able to set THIRD_PARTY_PACKAGE somewhere in mid of imports sequence by specifying regex and some imports without any pattern which Check will force to put at the end of imports list.

4) Removed that update as it really brings only mess.

romani commented:
all changes are good, 

1) please update commit comment to have "thirdparty package meaning is adjusted" as all commit messages appear in "What is new" on our website.

2) please confirm that where is not changes in violations on guava project with config for this Check from google_style.xml

alexkravin commented:
1) Done

2) 

Former version: http://alexkravin.github.io/CustomImportOrder/former/guava/checkstyle.html
Current: http://alexkravin.github.io/CustomImportOrder/guava/checkstyle.html

I see no difference, first config was taken from google_style before changes, second one - after.
